source/extensions/filters/network/thrift_proxy/buffer_helper.cc

```diff
@@ -0,0 +1,200 @@
+#include "extensions/filters/network/thrift_proxy/buffer_helper.h"
+
+#include "common/common/byte_order.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ThriftProxy {
+
+int8_t BufferHelper::peekI8(Buffer::Instance& buffer, size_t offset) {
+  ASSERT(buffer.length() >= 1);
+  int8_t i;
+  buffer.copyOut(offset, 1, &i);
+  return i;
+}
+
+int16_t BufferHelper::peekI16(Buffer::Instance& buffer, size_t offset) {
+  ASSERT(buffer.length() >= 2);
+  int16_t i;
+  buffer.copyOut(offset, 2, &i);
+  return be16toh(i);
+}
+
+int32_t BufferHelper::peekI32(Buffer::Instance& buffer, size_t offset) {
+  ASSERT(buffer.length() >= 4);
+  int32_t i;
+  buffer.copyOut(offset, 4, &i);
+  return be32toh(i);
+}
+
+int64_t BufferHelper::peekI64(Buffer::Instance& buffer, size_t offset) {
+  ASSERT(buffer.length() >= 8);
+  int64_t i;
+  buffer.copyOut(offset, 8, &i);
+  return be64toh(i);
+}
+
+uint16_t BufferHelper::peekU16(Buffer::Instance& buffer, size_t offset) {
+  ASSERT(buffer.length() >= 2);
+  uint16_t i;
+  buffer.copyOut(offset, 2, &i);
+  return be16toh(i);
+}
+
+uint32_t BufferHelper::peekU32(Buffer::Instance& buffer, size_t offset) {
+  ASSERT(buffer.length() >= 4);
+  uint32_t i;
+  buffer.copyOut(offset, 4, &i);
+  return be32toh(i);
+}
+
+uint64_t BufferHelper::peekU64(Buffer::Instance& buffer, size_t offset) {
+  ASSERT(buffer.length() >= 8);
+  uint64_t i;
+  buffer.copyOut(offset, 8, &i);
+  return be64toh(i);
+}
+
+int8_t BufferHelper::drainI8(Buffer::Instance& buffer) {
+  int8_t i = peekI8(buffer);
+  buffer.drain(1);
+  return i;
+}
+
+int16_t BufferHelper::drainI16(Buffer::Instance& buffer) {
+  int16_t i = peekI16(buffer);
+  buffer.drain(2);
+  return i;
+}
+
+int32_t BufferHelper::drainI32(Buffer::Instance& buffer) {
+  int32_t i = peekI32(buffer);
+  buffer.drain(4);
+  return i;
+}
+
+int64_t BufferHelper::drainI64(Buffer::Instance& buffer) {
+  int64_t i = peekI64(buffer);
+  buffer.drain(8);
+  return i;
+}
+
+uint32_t BufferHelper::drainU32(Buffer::Instance& buffer) {
+  uint32_t i = peekU32(buffer);
+  buffer.drain(4);
+  return i;
+}
+
+uint64_t BufferHelper::drainU64(Buffer::Instance& buffer) {
+  uint64_t i = peekU64(buffer);
+  buffer.drain(8);
+  return i;
+}
+
+double BufferHelper::drainDouble(Buffer::Instance& buffer) {
+  ASSERT(sizeof(double) == sizeof(uint64_t));
```

Comment:
`static_assert`

Reply:
Also, for the `union` trick below to work, `double` and `uint64_t` have to have the same endianness. It’s probably possible to test for that at compile time, but off the top of my head I can’t think of a simple way to do it.

Reply:
It's possible to do this at compile time in C with a trick similar to the union used below, but not in C++.

In fact, according to what I just read (https://stackoverflow.com/a/24431286 and https://stackoverflow.com/questions/11373203/accessing-inactive-union-member-and-undefined-behavior/11996970) the union below shouldn't be allowed either (though both Apache Thrift and Protobuf use it).